Documentation Changes for Contributing.MD (#1240)
microsoft/vscode-dev-containers: CONTRIBUTING.md

This document outlines a number of ways you can get involved.

> - [Contributing](#contributing)

Reviewer: Nice addition.

Finally, commit your changes and submit a PR - we'll take a look at it, provide any needed feedback, and then merge it in. We appreciate any and all feedback!

### Adding a Database Definition to an existing Container

> VS Code Dev Containers allow fairly straightforward modification for allowing multiple containers to run together. This is useful when adding database support to an existing container.

Reviewer 2: Maybe add something: ".. to run together leveraging `docker-compose`"

2. Create a new `docker-compose.yml` file.
   1. Add the existing `Dockerfile` to the `docker-compose` as an `app`. This will be the Dev Container.
   2. Add the definition for database to the `docker-compose`, version tagging it or using the `latest`. We don't use custom images for databases, as this is generally not needed.
   3. It will likely be necessary to network the containers so that the `app` can connect to the `db`. This can normally be done by adding the following to the `app` definition:

    ```yaml
        network_mode: service:db
    ```

3. Modify the `devcontainer.json` with any launch options or VS Code specific `settings.json` options that may be needed. 
   1. For example, when working with a SQL database, it might be necessary to modify the settings with something like the following:
   ```json
   		"sqltools.connections": [
			{
			  "name": "Container database",
			  "driver": "PostgreSQL",
			  "previewLimit": 50,
			  "server": "localhost",
			  "port": 5432,
			  // NOTE: database/username/password should match values in docker.compose.yml
			  "database": "postgres",
			  "username": "postgres",
			  "password": "postgres"
			}
		  ],
    ```

>     2. It might be necessary to add additional VS Code Extensions to allow for an easier dev experience.

Reviewer 2: Worth to mention that it depends of the database (i.e mssql vs postgres)

> 4. Modify the existing Tests or existing Application to ensure that the following criteria are met:

Reviewer 2: I'll try to force the suggestion to be a test rather than the app. Since CI would cover test if test.sh is present.

>     1. The application can successfuly ping the DB Container.

Reviewer 2: Forgot to mention that it is important to seed the db first.

Finally, commit your changes and submit a PR - we'll take a look at it, provide any needed feedback, and then merge it in. We appreciate any and all feedback!

### Adding a Database Definition to an existing Container

VS Code Dev Containers allow multiple containers to run together. This is useful, for example, when adding database support to an existing container, and can be leveraged with `docker compose.`

1. Create a new copy of an existing language definition folder, and hyphenate the database name.
   1. For example, `java` becomes `java-postgres`.
2. Create a new `docker-compose.yml` file.

>    1. Add the existing `Dockerfile` to the `docker-compose` as an `app`. This will be the Dev Container where your code gets executed.

Author: Done, thank you!

>    2. Add the definition for the database to the `docker-compose`, version tagging it or using the `latest`. Most usage scenarios don't require custom images for the database.

Author: Done, I called this out explicitly and rephrased this section.

>    2. Add the definition for database to the `docker-compose`, version tagging it or using the `latest`. We don't use custom images for databases, as this is generally not needed.

Reviewer 3: ```suggestion
   2. Add the definition for the database to the `docker-compose`, version tagging it or using the `latest`. Most usage scenarios don't require custom images for the database.
```

>    3. It will likely be necessary to network the containers so that the `app` can connect to the `db`. This can normally be done by adding the following to the `app` definition:

Reviewer 3: ```suggestion
   3. It's common that containers will need to be connected through a network so that the `app` can connect to the `db`. This can normally be done by adding the following to the `app` definition:
```

    ```yaml
        network_mode: service:db
    ```

3. Modify the `devcontainer.json` with any launch options or VS Code specific `settings.json` options that may be needed. 
   1. For example, when working with a SQL database, it might be necessary to modify the settings with something like the following:
   ```json

>    		"sqltools.connections": [

Reviewer: would it be worth poinging out that in this example, the settings are specific to the SQL Tools VS Code extension?

			{
			  "name": "Container database",
			  "driver": "PostgreSQL",
			  "previewLimit": 50,
			  "server": "localhost",
			  "port": 5432,
			  // NOTE: database/username/password should match values in docker.compose.yml
			  "database": "postgres",
			  "username": "postgres",
			  "password": "postgres"
			}
		  ],
    ```
    2. It might be necessary to add additional VS Code Extensions to allow for an easier dev experience. This could be database dependent (`MSSQL`, `NoSQL`, etc.)

> 4. Modify the existing Tests or add new tests to ensure that the following criteria are met:

Reviewer 3: ```suggestion
4. Modify the existing Tests or add new tests to ensure that the following scenarios are possible:
```

    1. The application can successfully ping the DB Container.
    2. The application can login to the DB Container using the default login data. 

>     3. The application can send a query to the existing DB and receive data from the DB. (Typically listing the databases in the cluster is considered a valid test.)

Reviewer: are we deploying database clusters? The examples we had were single instance.

Author: Apologies, I got confused because of the naming that PostgreSQL uses. I'll make this more clear.
